nexus/src/app/background/common.rs

```diff
@@ -0,0 +1,854 @@
+//!   to do) _and_ periodically (to deal with all manner of transient failures)
+//! * The reconciler's activity is idempotent: given the same underlying state
+//!   (e.g., database state), it always attempts to do the same thing.
+//! * Each activation of the reconciler accepts no input.  That is, even when we
+//!   think we know what changed, we do not use that information.  This ensures
+//!   that the reconciler really is idempotent and its actions are based solely
+//!   on the state that it's watching.  Put differently: having reconcilers
+//!   accept an explicit hint about what changed (and then doing something
+//!   differently based on that) bifurcates the code: there's the common case
+//!   where that hint is available and the rarely-exercised case when it's not
+//!   (e.g., because Nexus crashed and it's the subsequent periodic activation
+//!   that's propagating this change).  This is what we're trying to avoid.
+//! * We do allow reconcilers to be triggered by a `tokio::sync::watch` channel
+//!   -- but again, not using the _data_ from that channel.  There are two big
+//!   advantages here: (1) reduced latency from when a change is made to when
+//!   the reconciler applies it, and (2) (arguably another way to say the same
+//!   thing) we can space out the periodic activations much further, knowing
+//!   that most of the time we're not increasing latency by doing this.  This
+//!   compromises the "constant-work" pattern a bit: we might wind up running
+//!   the reconciler more often during busy times than during idle times, and we
+//!   could find that overloads something.  However, the _operation_ of the
+//!   reconciler can still be constant work, and there's no more than that
+//!   amount of work going on at any given time.
+//!
+//!   `watch` channels are a convenient primitive here because they only store
+//!   one value.  With a little care, we can ensure that the writer never blocks
+//!   and the readers can all see the latest value.  (By design, reconcilers
+//!   generally only care about the latest state of something, not any
+//!   intermediate states.)  We don't have to worry about an unbounded queue, or
+//!   handling a full queue, or other forms of backpressure.
+
+use chrono::DateTime;
+use chrono::Utc;
+use futures::future::BoxFuture;
+use futures::stream::FuturesUnordered;
+use futures::FutureExt;
+use futures::StreamExt;
+use nexus_db_queries::context::OpContext;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::Duration;
+use std::time::Instant;
+use tokio::sync::watch;
+use tokio::sync::Notify;
+use tokio::time::MissedTickBehavior;
+
+/// An operation activated both periodically and by an explicit signal
+///
+/// See module-level documentation for details.
+pub trait BackgroundTask: Send + Sync {
+    fn activate<'a, 'b, 'c>(
+        &'a mut self,
+        opctx: &'b OpContext,
+    ) -> BoxFuture<'c, serde_json::Value>
```

Comment:
Okay, after reading through the rest of the code, I ironically have found the **opposite** to be true.

It seems like a really common pattern is:
- Tasks exist in-memory, and may be executed periodically
- Tasks often have dependencies
- Tasks often communicate using the [tokio watch](https://docs.rs/tokio/latest/tokio/sync/watch/index.html) primitive, as a way to grab "whatever was the last value we saw from the prior task" to avoid repeating unnecessary work

I wonder if we want to make this output value a bit more first-class, and have tasks emit values that can be watched by downstream consumers? 

I don't actually think we should implement this now, but it's food for thought as we get more consumers of the API

Reply:
Yeah, good points.  It's pretty subtle trying to distinguish between "inputs that represent hints/shortcuts about what we're supposed to do" (which we're trying to avoid) and "inputs that represent the full canonical state, which just happens to have come from another task instead of the database [just because it's cleaner to factor things that way]".